plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageSettings.java

```diff
@@ -148,14 +151,45 @@ private static String buildConnectString(String account, @Nullable String key, @
         if (hasKey) {
             connectionStringBuilder.append(";AccountKey=").append(key);
         } else {
-            connectionStringBuilder.append(";SharedAccessSignature=").append(sasToken);
+            connectionStringBuilder.append(";SharedAccessSignature=").append(escapeSasToken(sasToken));
         }
         if (Strings.hasText(endpointSuffix)) {
             connectionStringBuilder.append(";EndpointSuffix=").append(endpointSuffix);
         }
         return connectionStringBuilder.toString();
     }
 
+    private static String escapeSasToken(String sasToken) {
+        StringBuilder escapedSasToken = new StringBuilder();
+        final String[] queryParams = sasToken.split("&");
+        for (int i = 0; i < queryParams.length; i++) {
+            String unescapedQueryParam = queryParams[i];
+            final String[] keyValue = unescapedQueryParam.split("=");
+            if (keyValue.length != 2) {
+                throw new IllegalArgumentException("Unable to parse sas token");
+            }
+
+            final String key = keyValue[0];
+            final String val = keyValue[1];
+            escapedSasToken.append(isAlreadyEncoded(key) ? key : encodeQueryParam(key));
+            escapedSasToken.append("=");
+            escapedSasToken.append(isAlreadyEncoded(val) ? val : encodeQueryParam(val));
+
+            if (i < queryParams.length - 1) {
+                escapedSasToken.append("&");
+            }
+        }
+
+        return escapedSasToken.toString();
+    }
+
+    private static boolean isAlreadyEncoded(String s) {
```

Comment:
I couldn't find another way to tell if a query param is already encoded or not, but we should do this check to avoid encoding twice an already encoded param (i.e. the signature parameter)